### ISSUE TITLE
AnalogOut (PWM) should be available on all digital enabled pins

### ISSUE THREAD

**Issue by finneyj** [lancaster-university/microbit-dal#115]
is currently restricted to those with an AtoD converter - not needed for PWM operation.

**Comment by sh1boot** [comment]
What actually needs to be done for this?  I'm currently using PWM output on pins 15 and 16 by simply bypassing MicroBitIO.cpp and instantiating the pins with `PIN_CAPABILITY_AD`; which is a lie, but it gets me past the corresponding test in `MicroBitPin::setAnalogValue()`.

Unless there are pins which cannot support PWM output, isn't it simply a matter of deleting that test?

**Comment by jaustin** [comment]
@finneyj I'm guessing that fixing this needs a bit of work to decouple PWM out from the pins that can do ADC when configured as input, is that right? 

If so, should we sketch out a design and mark as up for grabs?

**Comment by jamesadevine** [comment]
@jaustin Yes, there needs to be a separation between Analog generally, and specific Analog operations.

The enum currently looks like this:

```cpp
enum PinCapability{
    PIN_CAPABILITY_DIGITAL = 0x01,
    PIN_CAPABILITY_ANALOG = 0x02,
    PIN_CAPABILITY_AD = PIN_CAPABILITY_DIGITAL | PIN_CAPABILITY_ANALOG,
    PIN_CAPABILITY_ALL = PIN_CAPABILITY_DIGITAL | PIN_CAPABILITY_ANALOG
};
```
Although unnecessary, the perfect patch would be to bring out IN/OUT for each mode:

```cpp
enum PinCapability{
    PIN_CAPABILITY_DIGITAL_IN = 0x01,
    PIN_CAPABILITY_DIGITAL_OUT = 0x02,
    PIN_CAPABILITY_DIGITAL = PIN_CAPABILITY_DIGITAL_IN | PIN_CAPABILITY_DIGITAL_OUT,
    PIN_CAPABILITY_ANALOG_IN = 0x04,
    PIN_CAPABILITY_ANALOG_OUT = 0x08,
    PIN_CAPABILITY_ANALOG = PIN_CAPABILITY_ANALOG_IN | PIN_CAPABILITY_ANALOG_OUT,
    PIN_CAPABILITY_ALL = PIN_CAPABILITY_DIGITAL | PIN_CAPABILITY_ANALOG
};
```
I think this allows for maximised capability, you could thereby specify pins as Digital or Analog only, or any combination of the above.

The logic would remain simple in `MicroBitPin` as now you would only have to check against the flag that specifically relates the desired operation:

```cpp
int MicroBitPin::setServoPulseUs(int pulseWidth)
{
    //check if this pin has an analogue mode...
    if(!(PIN_CAPABILITY_ANALOG_OUT & capability))
        return MICROBIT_NOT_SUPPORTED;

    .....
```

**Comment by whaleygeek** [comment]
Related PXT ticket: https://github.com/Microsoft/pxt/issues/816

**Comment by whaleygeek** [comment]
Note, this seems to still not be working in PXT, Peli has directed us over to here:

https://github.com/Microsoft/pxt/issues/611

**Comment by whaleygeek** [comment]
![screen shot 2017-02-11 at 17 37 34](https://cloud.githubusercontent.com/assets/4086785/22855984/fdc5833c-f080-11e6-966f-f007b44be056.png)

Which one of these numbers is the DAL version?

**Comment by pelikhan** [comment]
For that, you need to look at the pxtarget.json file in PXT-microbit repo for that release.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by whaleygeek** [comment]
@pelikhan Which bit of this file please?

![screen shot 2017-02-11 at 17 43 29](https://cloud.githubusercontent.com/assets/4086785/22856023/a811d660-f081-11e6-99f8-79f7b5e36861.png)

**Comment by whaleygeek** [comment]
@pelikhan Oh, hang on? You guys have two repos? We only log issues against the core PXT but the code is actually built from pxt-microbit?

https://github.com/Microsoft/pxt-microbit/blob/master/pxtarget.json

**Comment by whaleygeek** [comment]
So, this is the link I presume? Thanks.

![screen shot 2017-02-11 at 17 48 06](https://cloud.githubusercontent.com/assets/4086785/22856065/4b77c4d6-f082-11e6-9032-e91ad6ec8024.png)

**Comment by pelikhan** [comment]
Yup v2.0.0-rc7 I believe this is the latest release.

Get Outlook for iOS<https://aka.ms/o0ukef>

_____________________________

**Comment by luniki** [comment]
@sh1boot Could you provide an example how you are using PWM output on pins 15 and 16? I would like to try this on my own too.

**Comment by jamesadevine** [comment]
@luniki You could test my PR: #264

**Comment by DavidWhaleMEF** [comment]
Is there any progress on this at all do you know? It's holding up a bug fix in PXT that we've got a long queue of people waiting for - ta!

https://github.com/Microsoft/pxt/issues/611

**Comment by jaustin** [comment]
And it's a change to make in DAL before RC...

J

**Comment by finneyj** [comment]
will do today

**Comment by finneyj** [comment]
Quick update.  All changes made. Just want to run a final test suite when i'm a bit less tired in the morning... I'll tag a new release early Wednesday AM UK time.

**Comment by Jaqster** [comment]
Hi guys.  We're releasing RC today.  We can work in the DAL changes after RC (it's not a breaking/UI change).

**Comment by jamesadevine** [comment]
@Jaqster The fix for this bug is in `v2.0.0-rc8` all tagged and ready to go :smile:
